rfcs/0075-expr-opaque-context.md

```diff
@@ -0,0 +1,161 @@
+3. Override some context variables is a little hard. We have to modify the variable, and don't forget to recover it when exiting context.
+
+The key point is that the call stack is in the ABA form: `frontend::run_query` -> `executor::exec` -> `expr::eval` -> `frontend::current_schema`
+
+And the `context` should be opaque for `executor` and `expr`, but concrate for `frontend`.
+
+## Design
+
+We'll utilize the [`provide_any`](https://github.com/nrc/provide-any) crate, although it's rejected by std.
+
+In `expr`:
+
+```rust
+pub trait Context: Provider {}
+// We can also provide some utilities such as `ChainedContext` here.
+pub trait Expression {
+  fn eval(&self, ctx: &dyn Context>, chunk: &DataChunk);
+  // ...
+}
+```
+
+In `frontend`:
+
+```rust
+pub struct FrontendContext {
+  meta_client: MetaClient,
+  session_variables: SessionVariables,
+  catalog: Catalog,
+  tz: TimeZone,
+}
+impl Provider for FrontendContext {
+  fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    demand
+      .provide_ref::<MetaClient>(&self.meta_client)
+      .provide_ref::<SessionVariables>(&self.session_variables)
+      .provide_ref::<Catalog>(&self.catalog)
+      .provide_value::<Tz>(self.tz);
+  }
+}
+impl Context for FrontendContext {}
+```
+
+In `frontend/expr`:
+
+```rust
+struct PgGetUserByIdExpression;
+// TODO: we can simplify the codes by another #[function] macro.
+impl expr::Expression for PgGetUserByIdExpression {
+  fn eval(&self, ctx: &dyn Context>, chunk: &DataChunk) {
+    let Ok(catalog) = request_ref::<Catalog>(ctx) else {
+      return Err("Not supported in the current context");
+    };
+    // retrieve the ids from chunk.
+    Ok(res)
+  }
+}
+// TODO: We should also consider how to construct the expression without depending on frontend.
+```
+
+In `executor`:
+
+```rust
+pub struct BackendContext {
+  tz: TimeZone,
+}
+impl Provider for BackendContext {
+  fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    // We can only provide a subset of context.
+    demand
+      .provide_value::<Tz>(self.tz);
+  }
+}
+```
+
+We can also dispatch some different logics based on the context:
+
+```rust
+pub trait Reporter {}
+// In frontend
+pub struct NoticeReporter;
+impl Reporter for NoticeReporter {}
+demand.provide_ref<dyn Reporter>(&self.notice_reporter);
+// In backend
+pub struct ErrorTableReporter;
+impl Reporter for ErrorTableReporter {}
+demand.provide_ref<dyn Reporter>(&self.error_table_reporter);
+```
+
+## Related work
+
+The pattern was widely used in golang. `context` is a part of std lib.
+
+There are also a hook [`useContext`](https://react.dev/reference/react/useContext) in react, which is excatly the same pattern as the RFC proposed.
+
+## Alternatives
+
+### task_local
+
+tokio's [`task_local` macro](https://docs.rs/tokio/latest/tokio/macro.task_local.html) can achieve the same result even simpler, and the only cost is the extensive use of global variables.
```

Comment:
`task_local` is not that "global" as the calling stack or the scope of evaluating an expression tree is quite clear. I would vote for this approach as it introduces minimal changes to the codebase.

Reply:
+1 for this approach. Adding `context` to the `eval` interface will make things more complicated and prevent easy use for most of the pure functions. Every time we call `eval`, we have to find out a context to pass in...

Reply:
You can even pass a () as context here. I don’t think it’s a heavy problem.

The key concept is pass it explicitly.

Reply:
We can also add a eval_with_empty_context.

It’s really simple.

Reply:
In fact, all `eval` needs a context and I'd prefer pass it explicitly instead of the one from unknown scope.

Reply:
Yes, my point is explicit context will put more burden on developers and make it error-prone. It's possible that developers forget to pass the context down through. This happened in our batch module, where we forgot to pass the shutdown channel which results in batch query not cancellable.

Another example is the `Future` API. Even if it has an explicit context in the `poll` method, most of the time the context is passed implicitly via `await`. 🤔

Reply:
I don't want to depend on the call stack heavily. If the code path changed, the code may be broken without any warnings.

Reply:
Well, I guess we can have an `eval_with_context` on the root and set the task-local variable there. In this way, theses local keys can be private in the `expr` crate and we can find where the context is set by listing all references of this function.

Reply:
Another solution: is it possible to put the context into the expression itself? I've discussed it with @shanicky yesterday.

Reply:
> Well, I guess we can have an eval_with_context on the root and set the task-local variable there

If you've already got the context at the expr root, it's really easy to just pass them recursively. It also gives you opportunity to override some variables with tracking.

The most of the work is, how to pass these variables from frontend to executors then expressions.

Reply:
> Another solution: is it possible to put the context into the expressions themselves? I've discussed it with @shanicky yesterday.

The lifetime of the context will be longer, it's not friendly for MetaCache, which is actually a ReadGuard and hold the read lock.

I also fully discussed with @fuyufjh last week.

Reply:
> Well, I guess we can have an eval_with_context on the root and set the task-local variable there.

Another issue is that our original intention was to make the expression module unaware of the values and types of these variables, but you completely deviated from this purpose.

Reply:
> > Well, I guess we can have an eval_with_context on the root and set the task-local variable there.
> 
> Another issue is that our original intention was to make the expression module unaware of the values and types of these variables (likely defined in frontend), but you completely deviated from this purpose.

In fact, these functions can be defined in the frontend module. The `#[function]` macro will register it into a global registry at start time. 🤡

Reply:
> Another issue is that our original intention was to make the expression module unaware of the values and types of these variables (likely defined in frontend), but you completely deviated from this purpose.

This sounds impossible. Although the context struct could be opaque, the type of each field must be accessible by the `expr` crate for expressions to use them.

Reply:
> > Another issue is that our original intention was to make the expression module unaware of the values and types of these variables (likely defined in frontend), but you completely deviated from this purpose.
> 
> This sounds impossible. Although the context struct could be opaque, the type of each field must be accessible by the `expr` crate for expressions to use them. Given that every field's type is known, whether the context itself should be concrete or opaque won't be a big deal.

Yes, so it seems that we **should and only can** define these function in the frontend.

Reply:
Yes, but we need to move the expression construction out of the expr crate.

Reply:
Besides, from my understanding, @TennyZhuang's point to introduce `provide` instead of using a simple `struct ExprContext` is he wants to make it more flexible. For example, if `BackendContext` exists, then the `eval()` function can get it; otherwise, it gets `None`.  So if you are sure that no context is needed somewhere (I don't know what can be the case... but just saying), it's possible to use an empty context.

Reply:
> Yes, but we need to move the expression construction out of the expr crate.

Not necessary? We can pass the opaque context into the constructor function.

Reply:
> I think evaluating expression doesn't happen everywhere.

It does happen everywhere in unit tests. And we have to provide an empty context for each of them. However, I think providing a `eval_with_default_context` method is acceptable.

Reply:
> It does happen everywhere in unit tests. 

Well. Indeed. 😄

If it costs too many changes to replace `eval` to `eval_with_default_context`, scoped task_local might look a little bit better. Either way LGTM.

By the way, it's worth mentioning that using `provide` or not is orthogonal to whether to put context in function call arguments or `task_local`: you may also put the `Context: Provider` as a task-local variable.